### PR TITLE
MODULES-1678 - Add show_diff attribute to concat and concat::fragment defined types

### PR DESCRIPTION
**Pull request by jdkindy**
Add show_diff to concat and concat::fragment types.

### CODE REVIEW

**Comment by HelenCampbell** [comment]
Hi,
Thank you for your work with this so far, this seems like a sensible addition to this module. However you would need to have all unit tests passing for us to have a look at this properly.

**Comment by jdkindy** [comment]
@HelenCampbell - I fixed the tests.  Please take a look to see if there is anything else I need to fix.

**Comment by jdkindy** [comment]
Found and removed a whitespace change.

**Comment by jdkindy** [comment]
@HelenCampbell Do I need to submit a new pull request?

File: Gemfile [hunk 1]
```diff
@@ -16,6 +16,7 @@ group :development, :unit_tests do
   gem 'simplecov',               :require => false
   gem 'puppet_facts',            :require => false
   gem 'json',                    :require => false
+  gem 'metadata-json-lint',      :require => false
```
Comment (igalic) [inline Gemfile]: is this not moduleynced?

Comment (jdkindy) [inline Gemfile]: I do not know what is meant by 'moduleynced'.  I had to add that line in order to get the tests to pass on my local box.

Comment (jdkindy) [inline Gemfile]: Removed line locally and retested successfully, so removed line from the commit.  Tests are successful.

**Comment by bmjen** [comment]
I think a `show_diff` param would be quite useful for concat, however I am not a fan of adding the corresponding param to the `concat::fragment`, since the fragment files themselves are supposed to be ingredients to the resulting file vs independently managed file resources. Also, when we do reintroduce concat as a native type, the utility of having a `show_diff` on fragments will become obsolete.

**Comment by jdkindy** [comment]
The reason for adding the show_diff parameter to concat::fragment resources is that diffs of the file resources defined as part of concat::fragment are included in the reports.  Leaving show_diff out of concat::fragment allows secrets to appear in reports.  Puppet runs show diffs of both the resultant concat file resource as well as diffs of the concat::fragment file resources.

**Comment by Vincent--** [comment]
I'd like to have this feature as well (for the same "secret" reason ;-) )

**Comment by bmjen** [comment]
@jdkindy I think I understand the reticence towards show_diff and how it corresponds to the fragment files. However, I am still against the exposure of the show_diff param in the fragment resource, as fragments should be internal resources only. I think an appropriate middle ground would be to remove the show_diff param from the fragment.. and force the fragment file to show_diff=false all the time. 

Thoughts?

**Comment by Vincent--** [comment]
I totally agree ! :-)

**Comment by jdkindy** [comment]
@bmjen I intended for the users to be able to choose fragments if they wished.  However, if the plan is for concat to move to a native type and the concat::fragment type will not have a show_diff, I am ok with setting it to false for all concat::fragment resources.

File: manifests/fragment.pp [hunk 2]
```diff
@@ -26,14 +26,14 @@
...
     $target,
-    $content = undef,
-    $source  = undef,
-    $order   = '10',
-    $ensure  = undef,
-    $mode    = undef,
-    $owner   = undef,
-    $group   = undef,
-    $backup  = undef
+    $content   = undef,
```
Comment (bmjen) [inline manifests/fragment.pp]: The alignment here needs to be fixed since `show_diff` was removed.

File: README.md [hunk 3]
```diff
@@ -246,6 +250,10 @@ Specifies whether the fragment should be included in the destination file or dis
 
 Reorders your fragments within the destination file. Fragments that share the same order number are ordered by name. Valid options: a string (recommended) or an integer. Default value: '10'.
 
+#####`show_diff`
```
Comment (bmjen) [inline README.md]: This section needs to be removed.

File: tests/fragment.pp [hunk 4]
```diff
@@ -1,19 +1,20 @@
...
+  group     => 'root',
+  mode      => '0664',
+  show_diff => true,
 }
 
 concat::fragment { '1':
-  target  => 'testconcat',
-  content => '1',
-  order   => '01',
+  target    => 'testconcat',
```
Comment (bmjen) [inline tests/fragment.pp]: alignment.

File: tests/fragment.pp [hunk 5]
```diff
@@ -1,19 +1,20 @@
...
+  content   => '1',
+  order     => '01',
 }
 
 concat::fragment { '2':
-  target  => 'testconcat',
-  content => '2',
-  order   => '02',
+  target    => 'testconcat',
```
Comment (bmjen) [inline tests/fragment.pp]: alignment.

**Comment by bmjen** [comment]
Thanks for the update @jdkindy! A few more comments and we're set for merge!

**Comment by jdkindy** [comment]
I think this commit takes care of all of the issues with the previous one.

**Comment by bmjen** [comment]
Awesome. Thanks for all the work @jdkindy !

**Comment by Vincent--** [comment]
Thks @jdkindy :-)

**Pull request merged**